api/storage_service.cc

```diff
@@ -403,6 +403,39 @@ void set_storage_service(http_context& ctx, routes& r) {
         });
     });
 
+    ss::repair_async_status_until_finished.set(r, [&ctx](std::unique_ptr<request> req) {
+        int id;
+        using clock = std::chrono::steady_clock;
+        clock::time_point expire;
+        try {
+            id = boost::lexical_cast<int>(req->get_query_param("id"));
+            // If timeout_ms is not provided, it means no timeout.
+            sstring s = req->get_query_param("timeout_ms");
+            int64_t timeout_ms = s.empty() ? int64_t(-1) : boost::lexical_cast<int64_t>(s);
+            if (timeout_ms < 0 && timeout_ms != -1) {
+                return make_exception_future<json::json_return_type>(
+                        httpd::bad_param_exception("timeout_ms can only be -1 (means no timeout) or non negative integer"));
+            }
+            if (timeout_ms < 0) {
+                expire = clock::time_point::max();
+            } else {
+                expire = clock::now() + std::chrono::milliseconds(timeout_ms);
+            }
+        } catch (std::exception& e) {
+            return make_exception_future<json::json_return_type>(httpd::bad_param_exception(e.what()));
+        }
+        return repair_get_status_until_finished(ctx.db, id, expire)
+                .then_wrapped([] (future<repair_status>&& fut) {
+            ss::ns_repair_async_status::return_type_wrapper res;
+            try {
+                res = fut.get0();
+            } catch (std::exception& e) {
+                return make_exception_future<json::json_return_type>(httpd::server_error_exception(e.what()));
+            }
+            return make_ready_future<json::json_return_type>(json::json_return_type(res));
+        });
+    });
```

Comment:
Suppose the API service is shut down. Will we have a use-after-free? Will it block shutdown?

Reply:
The repair service is shutodwn before http api service. 
When repair service shutdown(repair_shutdown()), it calls tracker::shutdown() which waits for _gate::close() and tracker::repair_await_completion runs under _gate. 
So after repair shutdowns, no pending api calls anymore.